rule/deep-exit.go

```diff
@@ -0,0 +1,85 @@
+package rule
+
+import (
+	"fmt"
+	"go/ast"
+
+	"github.com/mgechev/revive/lint"
+)
+
+// DeepExitRule lints program exit at functions other than main or init.
+type DeepExitRule struct{}
+
+// Apply applies the rule to given file.
+func (r *DeepExitRule) Apply(file *lint.File, arguments lint.Arguments) []lint.Failure {
+	var failures []lint.Failure
+	onFailure := func(failure lint.Failure) {
+		failures = append(failures, failure)
+	}
+
+	var exitFunctions = map[string]map[string]bool{
+		"os":      map[string]bool{"Exit": true},
+		"syscall": map[string]bool{"Exit": true},
+		"log": map[string]bool{
+			"Fatal":   true,
+			"Fatalf":  true,
+			"Fatalln": true,
+			"Panic":   true,
+			"Panicf":  true,
+			"Panicln": true,
+		},
+	}
+
+	w := lintDeepExit{onFailure, exitFunctions, false}
+	ast.Walk(w, file.AST)
+	return failures
+}
+
+// Name returns the rule name.
+func (r *DeepExitRule) Name() string {
+	return "deep-exit"
+}
+
+type lintDeepExit struct {
+	onFailure     func(lint.Failure)
+	exitFunctions map[string]map[string]bool
+	ignore        bool
+}
+
+func (w lintDeepExit) Visit(node ast.Node) ast.Visitor {
+	if stmt, ok := node.(*ast.FuncDecl); ok {
+		w.updateIgnore(stmt)
+		return w
+	}
+
+	if w.ignore {
+		return w
+	}
+
+	if se, ok := node.(*ast.ExprStmt); ok {
```

Comment:
It'll be easier to follow if it's:

```go
```
if se, ok := node.(*ast.ExprStmt); !ok {
  return nil
}
if ce, ok := se.X.(*ast.CallExpr); !ok {
  return nil
}
...

Reply:
:+1: